### PR TITLE
feat: support multiple routes per rollout

### LINKED ISSUE #33: Support multiple HTTPRoute resources per rollout
It's possible that the same rollout might need to change backend weights in multiple `HTTPRoute`/`TCPRoute` resources.
I'm running into a situation, where I have multiple `HTTPRoute` resources per service and can only change one.

This is similar to https://github.com/argoproj/argo-rollouts/issues/1100, which is about Istio `VirtualService` resources.

Example:
```yaml
plugins:
  argoproj-labs/gatewayAPI:
    httpRoutes: 
      - first
      - second
    namespace: default
```

If you think it makes sense I'm up for contributing.

### PR DESCRIPTION
**Pull request by svrakitin**
Closes #33

Introduces `httpRoutes` and `tcpRoutes` fields to plugin config, which allows us to reference multiple `HTTPRoute` and `TCPRoute` resources.

Given following routes:

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: first-route
spec:
  parentRefs:
    - name: gateway
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /first
    backendRefs:
    - name: argo-rollouts-stable-service
      kind: Service
      port: 80
    - name: argo-rollouts-canary-service
      kind: Service
      port: 80
---
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: second-route
spec:
  parentRefs:
    - name: gateway
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /second  
    backendRefs:
    - name: argo-rollouts-stable-service
      kind: Service
      port: 80
    - name: argo-rollouts-canary-service
      kind: Service
      port: 80
```

We can configure the plugin to update weights for each referenced route. For example:

```yaml
plugins:
  argoproj-labs/gatewayAPI:
    httpRoutes:
       - name: first-route
          useHeaderRoutes: true
       - name: second-route
```
**useHeaderRoutes** field indicates we will add header routes for this specific route or no

### CODE REVIEW

File: pkg/plugin/plugin.go
Comment (Philipp-Plotnikov) [inline pkg/plugin/plugin.go:59]: Lets create our own type in file types.go. Like
```go
type GatewayAPIRouteConfig struct {
  Kind string
  Name string
}
```

Comment (Philipp-Plotnikov) [reply pkg/plugin/plugin.go:59]: I think it is not good idea use type conversions constantly when we can not use them

File: pkg/plugin/plugin.go
Comment (Philipp-Plotnikov) [inline pkg/plugin/plugin.go:52]: Lets move as seperate line, if condition will be more readable
```go
err := json.Unmarshal(rollout.Spec.Strategy.Canary.TrafficRouting.Plugins[PluginName], gatewayAPIConfig)
```

File: pkg/plugin/plugin.go
Comment (Philipp-Plotnikov) [inline pkg/plugin/plugin.go:87]: Lets reuse gatewayAPIConfig, assign routeName to gatewayAPIConfig.HTTPRoute

**Comment by Philipp-Plotnikov** [comment]
@svrakitin please ping me when you will resolve all comments

**Comment by svrakitin** [comment]
@Philipp-Plotnikov Made some changes

File: pkg/plugin/plugin.go
Comment (Philipp-Plotnikov) [inline pkg/plugin/plugin.go:48]:
```go
json.Unmarshal(rollout.Spec.Strategy.Canary.TrafficRouting.Plugins[PluginName], &gatewayAPIConfig)
```

**Comment by Philipp-Plotnikov** [comment]
@svrakitin Dont you think that
'''yaml
plugins:
  argoproj-labs/gatewayAPI:
    httpRoutes: 
      - first
      - second
     tcpRoutes:
...
 '''
Will be easier to use ? And also in so case we dont need additional structure in GatewayAPIConfig, right ? We will have seperate fields
```go
HTTPRouteList []string 
TCPRouteList []string
```

**Comment by svrakitin** [comment]
I don't have a preference as long as it does the job. If you feel this makes more sense I'll update.

File: pkg/plugin/plugin.go
Comment (Philipp-Plotnikov) [inline pkg/plugin/plugin.go:77]: And in this case we dont need this check(if we will use httpRoutes and so on)

**Comment by Philipp-Plotnikov** [comment]
> I don't have a preference as long as it does the job. If you feel this makes more sense I'll update.

Then, please update as it seems easier to use and less code

File: pkg/plugin/plugin.go
Comment (Philipp-Plotnikov) [inline pkg/plugin/plugin.go:102]: remove empty line

**Comment by svrakitin** [comment]
Done. Not sure that using separate fields makes it easier though.

File: pkg/plugin/plugin.go
Comment (Philipp-Plotnikov) [inline pkg/plugin/plugin.go]: Have code duplication. Lets make seperate function for that
```go
forEachGatewayAPIRoute(listOfRoutes, func (name) error {
     return r.setHTTPRouteWeight(rollout, desiredWeight, additionalDestinations, GatewayAPITrafficRouting{
			Namespace: gatewayAPIConfig.Namespace,
			HTTPRoute: name,
    })
})
```
And below you can reuse it

File: pkg/plugin/plugin.go
Comment (Philipp-Plotnikov) [inline pkg/plugin/plugin.go:63]: Lets make here function for that
```go
if isGatewayAPIRouteListsEmpty(gatewayAPIConfig) {

}
```

File: pkg/plugin/plugin.go
Comment (svrakitin) [reply pkg/plugin/plugin.go]: Looks overkill, but I guess whatever it takes to get this done ;)

File: pkg/plugin/plugin.go
Comment (Philipp-Plotnikov) [inline pkg/plugin/plugin.go:78]: https://github.com/argoproj-labs/rollouts-plugin-trafficrouter-gatewayapi/pull/34#discussion_r1480056376
To be sure that it is seen, Github decided it is oudated

**Comment by Philipp-Plotnikov** [comment]
Please update description to the PR to understand how to test this feature for users who wants to test it. I think it is very useful feature and somebody wanted to test it not reading the code

**Comment by Philipp-Plotnikov** [comment]
> Done. Not sure that using separate fields makes it easier though.

Code became cleaner and tcpRoutes and httpRoutes will be seperated and easier to find needing route in manifest

File: pkg/plugin/plugin.go
Comment (Philipp-Plotnikov) [reply pkg/plugin/plugin.go]: For example one of the feature that will be released on this week
https://github.com/argoproj-labs/rollouts-plugin-trafficrouter-gatewayapi/pull/31

Comment (Philipp-Plotnikov) [reply pkg/plugin/plugin.go]: And Im sure it will grow )

**Comment by Philipp-Plotnikov** [comment]
When you will finish please update me to be sure that it is ok, but I think it was last comments. We will wait a little bit on this week to give users a chance to run it locally and test it. After I or Kostis merge this PR and ping you here with congratulations )

**Comment by svrakitin** [comment]
Alright, I hope that's it. Everything else can be addressed by other PRs as pure cosmetics. Thanks.

**Comment by Philipp-Plotnikov** [comment]
@svrakitin great job!

**Comment by supersyn** [comment]
Hi all! I've built the plugin using this branch locally (run the `make` command in the main folder), but when trying to use it I've stumbled upon the same issue as already reported [here](https://github.com/argoproj-labs/rollouts-plugin-trafficrouter-gatewayapi/issues/29):

`failed to get traffic router plugin argoproj-labs/gatewayAPI: unable to start plugin system: unable to get plugin client (argoproj-labs/gatewayAPI) for ping: exec: Stdout already set\n" error="<nil>"`

The official v0.0.0-rc1 version of the plugin works without problems.

go version used for local build: go1.20.14 darwin/arm64

[Built artifact](https://azure-to-aws-migration-test.s3.eu-central-1.amazonaws.com/gateway-api-plugin-linux-amd64) I am using for testing.

I tried to reset the Argo Rollouts Controller deployment multiple times, delete the rollout, and reapply it again, but nothing helped.

Do you maybe have any idea what could be the issue here?

Thank you!

**Comment by Philipp-Plotnikov** [comment]
Hi @supersyn ! It seems because some process have been already using needing resources. Im using Vscode and in development in some cases I need to kill process manually to free resources to start new version of plugin. Did you check it ?

**Comment by svrakitin** [comment]
Recent changes in upstream introduced too many conflicts and I would appreciate help with resolving those as I will only be able to pick this up after Feb 24.

**Comment by kostis-codefresh** [comment]
@svrakitin We will resolve the conflicts since we introduced them. Sorry for the confusion.

**Comment by Philipp-Plotnikov** [comment]
@svrakitin it will be great if you will test it locally to be sure it is ok

**Comment by Philipp-Plotnikov** [comment]
@kostis-codefresh it seems PR is ready. I tested locally, it works fine. It will be great if someone will test it locally too. 
To create local build, run in the root of repository:
```
make local-build
```
Will we add documentation in this PR ?

File: examples/traefik/README.md
Comment (Philipp-Plotnikov) [inline examples/traefik/README.md:345]: Will remove it, not true

**Comment by Philipp-Plotnikov** [comment]
@svrakitin I think we will merge this PR on the following week if anyone will find bugs

**Review by kostis-codefresh — APPROVED** [review]

**Comment by Philipp-Plotnikov** [comment]
@svrakitin our congratulations with contribution! Great job! Thank you!

**Comment by kostis-codefresh** [comment]
@svrakitin Documentation is now here https://rollouts-plugin-trafficrouter-gatewayapi.readthedocs.io/en/latest/features/multiple-routes/

Any feedback welcome

**Pull request merged**
